enhancements/local-storage/automatic-detection.md

```diff
@@ -0,0 +1,416 @@
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+	Spec              LocalVolumeDiscoverySpec   `json:"spec"`
+	Status            LocalVolumeDiscoveryStatus `json:"status,omitempty"`
+}
+
+type LocalVolumeDiscoverySpec struct {
+	// Nodes on which the automatic detection policies must run.
+	// +optional
+	NodeSelector *corev1.NodeSelector `json:"nodeSelector,omitempty"`
+	// If specified tolerations is the list of toleration that is passed to the
+	// LocalVolumeDiscovery Daemon
+	// +optional
+	Tolerations []corev1.Toleration `json:"tolerations,omitempty"`
+}
+
+type LocalVolumeDiscoveryStatus struct {
+	Phase string `json:"phase,omitempty"`
+}
+
+type NodeDeviceMap List struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+	Items           []NodeDeviceMap `json:"items"`
+}
+
+type NodeDeviceMap struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+	Spec              NodeDeviceMapSpec   `json:"spec"`
+	Status            NodeDeviceMapStatus `json:"status,omitempty"`
+}
+
+type NodeDeviceMapSpec struct {
+	NodeName string `json:"nodeName"`
+	// If specified tolerations is the list of toleration that is passed to the
+	// LocalVolumeDiscovery Daemon
+	// +optional
+	Tolerations []corev1.Toleration `json:"tolerations,omitempty"`
+}
+
+type NodeDeviceMapStatus struct {
+	Phase string `json:"phase,omitempty"`
+	// Blockdevices will be containing the output of following command unmarshalled into the structured
+	// "lsblk -o name,model,vendor,kname,pkname,type,uuid,tran,ro,rm,parttype,serial,size,rota,mountpoint"
+	Blockdevices []Device `json:"blockdevices"`
+}
+
+type Device struct {
+	Name       string   `json:"name,omitempty"`
+	Model      string   `json:"model,omitempty"`
+	Vendor     string   `json:"vendor,omitempty"`
+	Kname      string   `json:"kname,omitempty"`
+	Pkname     string   `json:"pkname,omitempty"`
+	Type       string   `json:"type,omitempty"`
+	UUID       string   `json:"uuid,omitempty"`
+	Tran       string   `json:"tran,omitempty"`
+	Ro         string   `json:"ro,omitempty"`
+	Parttype   string   `json:"parttype,omitempty"`
+	Serial     string   `json:"serial,omitempty"`
+	Size       string   `json:"size,omitempty"`
+	Mountpoint string   `json:"mountpoint,omitempty"`
+	Rm         string   `json:"rm,omitempty"`
+	Children   []Device `json:children,omitempty`
+}
+```
+
+#### Workflow of Device Discovery
+- The discovery of local devices will be done by the `LocalVolumeDiscovery` CR. Once the `LocalVolumeDiscovery` CR is deployed, the controller of the `LocalVolumeDiscovery` CR will be managing a `NodeDeviceMap` CR for each node that matched the `nodeSelector`. The `NodeDeviceMap` CR will be exposing the available local devices of the node in its status. These information will mostly be the output of  `lsblk -o name,model,vendor,kname,pkname,type,uuid,tran,ro,rm,parttype,serial,size,rota,mountpoint`
+
+Example of `LocalVolumeDiscovery` CR:
+```yaml
```

Comment:
What if multiple LocalVolumeDiscovery CRs are created with overlapping node selectors? Or what if the NodeDeviceMap CR already exists for another reason? Does a new NodeDeviceMap CR get created in order to ensure a new discovery is triggered on a node? Or would the existing CR remain? I would think triggering a new discovery makes sense, whether by updating the existing one or deleting the existing one and creating a fresh one.

Reply:
yeah we will retrigger a new discovery job whenever the CR is updated.

Reply:
Sounds good, is this captured in the doc somewhere?

Reply:
added it